Add ResponseStatusCodeRector, fixes #1229

shyim opened the pull request:
Rector writing makes fun 😅

shyim commented on packages/Symfony/src/Rector/Identical/ResponseStatusCodeRector.php:
```diff
@@ -0,0 +1,133 @@
+            return null;
+        }
+
+        $methodCall->args[0] = new Node\Arg(new Node\Expr\ClassConstFetch(new Node\Name('\\Symfony\\Component\\HttpFoundation\\Response'), new Node\Identifier($this->codeToConst[$statusCode->value])));
```
Hey @TomasVotruba, is there a way to add a use statement?

TomasVotruba replied:
Hi, some people preffer FQN, some import it. So that's job of coding standard. 

Also, you can use `$this->createClassConstant()` methodh here.

TomasVotruba commented on packages/Symfony/src/Rector/Identical/ResponseStatusCodeRector.php:
```diff
@@ -0,0 +1,132 @@
+
+    private function processMethodCall(Node\Expr\MethodCall $methodCall): ?Node\Expr\MethodCall
+    {
+        if (!$this->isType($methodCall->var, 'Symfony\\Component\\HttpFoundation\\Response') || $this->getName($methodCall)!== 'setStatusCode') {
```
Instead of `getName() ===` you can use `isName()`

TomasVotruba commented on packages/Symfony/src/Rector/Identical/ResponseStatusCodeRector.php:
```diff
@@ -0,0 +1,132 @@
+        $ref = new \ReflectionClass('Symfony\\Component\\HttpFoundation\\Response');
+
+        foreach ($ref->getConstants() as $key => $value) {
+            if (strpos($key, 'HTTP_') === 0) {
```
Instead of reflection, the Rector should work without depending on any `Response` class.

The best practise is to make specific property with map of values:

```php
private $codeToConstant = [
   200 => 'HTTP_SUCCESS'
   // ...
];
```


Then iterate it.

TomasVotruba commented on packages/Symfony/src/Rector/Identical/ResponseStatusCodeRector.php:
```diff
@@ -0,0 +1,132 @@
+     */
+    public function getNodeTypes(): array
+    {
+        return [Identical::class, Node\Expr\MethodCall::class];
```
What about `NotIdentical`? `NotEqual`?...

Better use `BinaryOp` to cover any compare.

TomasVotruba commented:
> Rector writing makes fun sweat_smile

I'm glad you enjoy the same way I do :D

shyim commented:
Fixed all your comments. Also, moved everything to BinaryOp folder 😊 
It would be helpful when somewhere is a doc about Helper Functions 🙈. I didn't know that functions thanks! 😅

TomasVotruba commented:
> Fixed all your comments. 

:+1: 

> It would be helpful when somewhere is a doc about Helper Functions :see_no_evil:

They're so many adn growing they would be outdated by next week :). The best is to `$this-><hit space>` in a class extending `AbstractRector`.

TomasVotruba commented on packages/Symfony/src/Rector/BinaryOp/ResponseStatusCodeRector.php:
```diff
@@ -0,0 +1,186 @@
+            return $this->processMethodCall($node);
+        }
+
+        if ($node instanceof Node\Expr\BinaryOp\Identical || $node instanceof Node\Expr\BinaryOp\NotIdentical || $node instanceof Node\Expr\BinaryOp\Equal || $node instanceof Node\Expr\BinaryOp\NotEqual) {
```
What about `>`, `<`, `=<`, `||`, `&&` ...?

TomasVotruba commented on packages/Symfony/src/Rector/BinaryOp/ResponseStatusCodeRector.php:
```diff
@@ -0,0 +1,186 @@
+            return null;
+        }
+
+        $methodCall->args[0] = new Node\Arg($this->createClassConstant('Symfony\\Component\\HttpFoundation\\Response', $this->codeToConst[$statusCode->value]));
```
Why there is `\\` in class name?

It's used on multiple classes, so beter extract is to `RESPONSE_CLASS` constant

TomasVotruba commented on packages/Symfony/src/Rector/BinaryOp/ResponseStatusCodeRector.php:
```diff
@@ -0,0 +1,186 @@
+
+        $node->right = $this->createClassConstant('Symfony\\Component\\HttpFoundation\\Response', $this->codeToConst[$node->right->value]);
+
+        return $node;
```
What about the other side?

```php
$response->getStatus() !== 200;
200 !== $response->getStatus();
```

TomasVotruba commented on packages/Symfony/tests/Rector/BinaryOp/ResponseStatusCodeRector/ResponseStatusCodeRectorTest.php:
```diff
@@ -0,0 +1,21 @@
+    {
+        return ResponseStatusCodeRector::class;
+    }
+}
```
Could you add 1-2 more tests? E.g.:

- with the other side match 
- code that is not in the list, e.g. `1241`

TomasVotruba commented on packages/Symfony/src/Rector/BinaryOp/ResponseStatusCodeRector.php:
```diff
@@ -0,0 +1,186 @@
+        /** @var Node\Expr\MethodCall $left */
+        $left = $node->left;
+
+        if (!$this->isType($left, 'Symfony\\Component\\HttpFoundation\\Response') || !$this->isName($left, 'getStatusCode')) {
```
Here we have the conventions to keep single case conditions:

```php
if (!$this->isType($left, 'Symfony\\Component\\HttpFoundation\\Response')) {
    return null;
} 

if (!$this->isName($left, 'getStatusCode')) {
    return null;
}
```

It makes reading easier and prevents from joining zillions of conditions on one line (or multiple lines).

TomasVotruba commented:
Also please do not force push until the PR is accepted. 
It's makes review easier to just check the new changes.

TomasVotruba commented on packages/Symfony/tests/Rector/BinaryOp/ResponseStatusCodeRector/Fixture/fixture.php.inc:
```diff
@@ -0,0 +1,33 @@
+    }
+}
+
+?>
```
Nice and simle test coverring all it needs with minimal code :+1:  

Great job! :clap:

TomasVotruba commented on packages/Symfony/tests/Rector/BinaryOp/ResponseStatusCodeRector/Fixture/fixture.php.inc:
```diff
@@ -27,6 +34,13 @@ class SomeController
+        if ($response->getStatusCode() < \Symfony\Component\HttpFoundation\Response::HTTP_OK) {}
+        if ($response->getStatusCode() != \Symfony\Component\HttpFoundation\Response::HTTP_OK) {}
+        if ($response->getStatusCode() !== \Symfony\Component\HttpFoundation\Response::HTTP_OK) {}
+        if ($response->getStatusCode() === \Symfony\Component\HttpFoundation\Response::HTTP_OK || $response->getStatusCode() === \Symfony\Component\HttpFoundation\Response::HTTP_CREATED)  {}
```
Nice :+1:

TomasVotruba commented:
What level should be in? Is it related to upgrade of specific Symfony version?

If not, we could create new `symfony-code-quality.yaml` level for features that improve Symfony code.

shyim commented:
I have created the config symfony-code-quality.yaml :)

TomasVotruba commented:
Thank you for your contribution!
